developer/develop.py

```diff
@@ -396,29 +429,49 @@ def _select_buildings(self, df, p, custom_selection_func):
             Index of buildings selected for development
 
         """
-
+        insufficient_units = df.net_units.sum() < self.target_units
         if custom_selection_func is not None:
             build_idx = custom_selection_func(self, df, p)
-        elif df.net_units.sum() < self.target_units:
+        elif insufficient_units & (not self.keep_suboptimal):
             print("WARNING THERE WERE NOT ENOUGH PROFITABLE UNITS TO",
                   "MATCH DEMAND")
             build_idx = df.index.values
         elif self.target_units <= 0:
             build_idx = []
         else:
-            # we don't know how many developments we will need, as they differ
-            # in net_units. If all developments have net_units of 1 than we
-            # need target_units of them. So we choose the smaller of available
-            # developments and target_units.
-            choices = np.random.choice(df.index.values,
-                                       size=min(len(df.index),
-                                                self.target_units),
-                                       replace=False, p=p)
-            tot_units = df.net_units.loc[choices].values.cumsum()
-            ind = int(np.searchsorted(tot_units, self.target_units,
-                                      side="left")) + 1
-            build_idx = choices[:ind]
-
+            if not self.keep_suboptimal:
+                # We don't know how many developments we will need, as they
+                # differ in net_units. If all developments have net_units of 1
+                # than we need target_units of them. So we choose the smaller
+                # of available developments and target_units.
+                choices = np.random.choice(df.index.values,
+                                           size=min(len(df.index),
+                                                    self.target_units),
+                                           replace=False, p=p)
+                tot_units = df.net_units.loc[choices].values.cumsum()
+                ind = int(np.searchsorted(tot_units, self.target_units,
+                                          side="left")) + 1
+                build_idx = choices[:ind]
+            else:
+                # If sub-optimal proposals are in the pool, then sort by
+                # profitability and only select a single proposal per parcel.
+                # Gaussian noise optionally (default noise_scale is 0) added
+                # to profits to allow possibility of less profitable proposals
+                # out-competing the most profitable.
+                noise = np.random.normal(0, self.noise_scale, len(df))
+                df['profit_noisy'] = df.max_profit + noise
+                df.sort_values('profit_noisy', ascending=False, inplace=True)
```

Comment:
Sorry for the late review. Why do you want to use `random noise and an absolute sort` instead of the `weighted random sort` that we used before? `np.random.choice` will shuffle the data with respect for the probabilities provided.